Usermod INA219 + INA226 Current, Voltage & Power Sensor

KrX3D opened the pull request:
Hi,

i created a usermod for the ina219 Current, Voltage & Power Sensor:

![INA219-Pin-Outs](https://github.com/user-attachments/assets/4fa56590-1298-4e4d-9d19-85ed64f70e94)

I took a liking to the INA226 Usermod here:

https://github.com/Aircoookie/WLED/tree/0_15/usermods/INA226_v2

and took some parts and added my 3 cents to it.

I hope the readme states everything needed to get what can be done.

the only think i couldnt solved is to add the sensors data to the wled home assistant integration page (if thats even possible) so instead it is added just to mqtt. if there is a way to have the autodiscover to add it directly to the wled integrations page, i would be more than happy to change the usermod.

also i hope i added everything correctly in git, since its my first time working like this in git


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - INA2xx support (INA219/INA226): per-sensor voltage, current, power, configurable sampling/precision, and energy tracking with daily/monthly/total persistence and resets.

- **Integrations**
  - MQTT publishing with retained-state restoration and Home Assistant discovery; per-sensor MQTT/state options and change-detection publishing.

- **Documentation**
  - Comprehensive README covering setup, wiring, configuration, build, and usage.

- **Chores**
  - Module registration and build manifest/dependency declaration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

KrX3D committed:
Update usermods_list.cpp

KrX3D committed:
Update const.h

KrX3D committed:
Add INA219 Usermod

KrX3D committed:
Update usermod_ina219.h

KrX3D committed:
Update usermod_ina219.h

Formatting

DedeHai commented:
since you keep adding commits, is this work in progress? if so, mark it as such or better yet: commit to this PR once you are finished.

KrX3D commented:
> since you keep adding commits, is this work in progress? if so, mark it as such or better yet: commit to this PR once you are finished.

Hi,

no its finsied. I just did some small formatting changes while waiting for a reply. But now im happy with it ;)

DedeHai commented:
ok nice. did you test it working on different ESPs?

KrX3D commented:
Hi, 

i tested it now also with an ESp32 Wroom-32 (esp32dev) and it also worked.

did i forget something so it will be added to version 0,15? 
im new to git with all this pull request stuff

KrX3D committed:
Update usermod_ina219.h

changed uint8_t to int8_t for _sdaPin and _sclPin --->  to be able to pick "unused" in pin selection which is -1,

KrX3D committed:
Update usermod_ina219.h

Several optimizations for WLED_MQTT_CONNECTED and if Usermod is enabled/disabled

KrX3D committed:
Merge branch 'main' into 0_15

KrX3D committed:
Update usermods_list.cpp

fixed 
  endif
to
  #endif

KrX3D committed:
indentation fix Part 1

KrX3D committed:
indentation part 2

KrX3D committed:
indentation fix

KrX3D commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,702 @@
+        #ifndef WLED_DISABLE_MQTT     
+            // Publish to MQTT if enabled
+            if (mqttPublish) {
+				if (mqttPublishAlways || hasValueChanged()) {
```
HI, im not really sure why it was wrong. Locally it was right and editing in git dev also showed it right, only on the normal view it was shown wrong.

i edited each line and moved it to the front and than I indented it, hopefully its now ok.

KrX3D committed:
use enum  from INA219_WE library

KrX3D commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,702 @@
+    }
+
+    // ADC mode enumeration
+    enum class ADCMode {
```
i overlooked it while porting this from my arduino code.

i changed it, hopefully its ok now. I tested it and it seems to work

KrX3D committed:
PinManager allocatePin SDA SCL

DedeHai commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,734 @@
+	#endif
+
+	// Decimal factor for current/power readings
+	#ifdef INA219_DECIMAL_FACTOR
```
all these defines could be more readable if you put them at the top of the file:
#ifndef SOME_CONST
#define SOME_CONST 5
#endif

and then just set the variable to SOME_CONST

KrX3D replied:
Hi, im not totally sure if i understand. you mean like this?:

```
#pragma once

// Define constants at the top
#ifndef INA219_ENABLED
    #define INA219_ENABLED false
#endif

#ifndef INA219_I2C_ADDRESS
    #define INA219_I2C_ADDRESS 0x40
#endif

......
......
......

#include "wled.h"
#include <Wire.h>
#include <INA219_WE.h>

class UsermodINA219 : public Usermod {
private:
    bool enabled = INA219_ENABLED;
    uint8_t _i2cAddress = INA219_I2C_ADDRESS;

	bool enabled = INA219_ENABLED;
	uint8_t _i2cAddress = INA219_I2C_ADDRESS;

```

DedeHai replied:
this is how its done throught the rest of WLED code.

KrX3D replied:
ok, hopefully done correctly:

https://github.com/Aircoookie/WLED/pull/4237/commits/8a142a72aee8a4ce92b3b1c4783f95782c5b3171

DedeHai commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,734 @@
+		bool haDiscoverySent = true;
+	#endif
+
+	// I2C SDA and SCL pins (default SDA = 8, SCL = 9 if not defined)
```
why use pin 8 and 9 and not the wled default pins? I also just noticed there is a INA226 UM already, what is the difference between that and the INA219 (hardware differences?) maybe it would be easier to combine the two into one as the ICs seem very similar in function?

KrX3D replied:
> why use pin 8 and 9 and not the wled default pins?

Which pins are the default? To be honest, I'm not entirely sure. I just took this part from other usermods where pins were also assigned this way. I read somewhere that on the ESP32, you can choose from many pins for SDA and SCL, so I wrote it this way and made it configurable so the user can select their own pins.

If that's not the case, how can I change it? I'm not completely sure.

> I also just noticed there is a INA226 UM already, what is the difference between that and the INA219 (hardware differences?) maybe it would be easier to combine the two into one as the ICs seem very similar in function?

The libraries used are different, and the functions vary because different registers and data formats are used. I didn't want to force everything together.

I'm not sure how to answer this—I was simply curious about using that IC and wanted to create a usermod since I was migrating some devices to WLED.

DedeHai replied:
wled defaults are configured in usermod settings but can be chosen to whatever pins are supported. so your UM would either define a second I2C (not sure the library or hardware supports that) or it will overrule the other definition or just throw an error. 

joining the other INA usermod is not necessary, but you could borrow from that code. if both ICs do the same (which I think they do more or less) you could extend the existing UM by making the IC type selectable in the settings. just an idea.

DedeHai replied:
I never used I2C in wled, but I assume your assessment is correct, just do it like other UM do it. WLED most likely takes care of the wire.begin() and it also allocates the pins, so you should be able to just use it without any initialization or pin allocation.

KrX3D replied:
ok i deleted everything sda/scl related i had and picking the pins at the usermod section -> Global I2C works now

https://github.com/Aircoookie/WLED/pull/4237/commits/a6fc7f43ff47cbde37107991bf103dc22bf0749b

KrX3D committed:
indentation

KrX3D commented on wled00/const.h:
```diff
@@ -205,6 +205,7 @@
 #define USERMOD_ID_PIXELS_DICE_TRAY      54     //Usermod "pixels_dice_tray.h"
 #define USERMOD_ID_DEEP_SLEEP            55     //Usermod "usermod_deep_sleep.h"
 #define USERMOD_ID_RF433                 56     //Usermod "usermod_v2_RF433.h"
+#define USERMOD_ID_INA219       		     57     //Usermod "usermod_ina219.h"
```
fixed, im totally not sure why this happens for me it looks always like this in github.dev:

![image](https://github.com/user-attachments/assets/13166777-ce3b-4e81-a62c-920cf7a9ed60)

KrX3D replied:
hope its now alright

KrX3D committed:
Refactor INA219 Usermod Configuration

- Moved configurable settings to the top of the file for better readability.
- Defined default values for each configuration option using #ifndef directives.
- Updated default I2C pins to WLED ESP32 defaults (SDA: 21, SCL: 22).
- Cleaned up variable initialization to use pre-defined constants for configuration.

KrX3D committed:
Changed SDA/SCL to Global I2C, managed by WLED

DedeHai commented:
looks good now.
did you test that the UM compiles on all platforms? if it does not: throw compile errors for platforms not supported.

KrX3D committed:
Updated pictures, since SDA/SCL pin selection through usermod was removed

KrX3D committed:
Delete usermods/INA219/img/usermod_settings.png

KrX3D committed:
Updated pictures, since SDA/SCL pin selection through usermod was removed

KrX3D committed:
Update ReadMe since SDA/SCL pin selection was removed via usermod

KrX3D committed:
Delete usermods/INA219/img/usermod_settings_pins_selection.PNG

KrX3D committed:
Fix PNG to png extension

KrX3D committed:
Usage info for sda/scl

KrX3D commented:
Hi,
i updated ReadMe and pictues since i removed the SDA/SCL pin selection from the usermod.

------------

im not sure if i need also should add under lib_deps inside platformio.ini:
    wollewald/INA219_WE@~1.3.8

i already added the info to the readme. I mean here:

![image](https://github.com/user-attachments/assets/5af2f485-fdc6-42ff-a5cd-b0315dbacc20)

---------------------

i compiled it for all default_envs (v0.15) and got no errors:

![image](https://github.com/user-attachments/assets/81b03be6-188a-481e-93ce-00547a1ce3fc)

--------------

for my own info, is there a way to exclude COM1 from  beeing uploaded? compiling was taking to long that way

KrX3D committed:
Update const.h

Solve for: This branch has conflicts that must be resolved

KrX3D committed:
Merge branch 'main' into 0_15

KrX3D committed:
Update usermods_list.cpp

Fix missing pound in #endif.

Line 504 has “endif” without a leading #. This is likely a small typographical mistake that can break compilation.

KrX3D committed:
Update usermod_ina219.h

Handle INA219 initialization failure more robustly.

Currently, you only log "INA219 initialization failed!" if _ina219->init() fails. Consider disabling enabled or providing fallback logic to avoid subsequent read attempts on a failed initialization, preventing usage of a potentially invalid object.

KrX3D committed:
Update usermods_list.cpp

coderabbitai ->

502-504: Consider moving the registration to maintain alphabetical order.

The registration code is correctly implemented but could be moved to maintain alphabetical order with other usermods (between INA226 and LD2410).

KrX3D committed:
coderabbitai imprevments ->

316-357: Add units to JSON keys for clarity.
You produce JSON keys like "current", "power", etc. Consider including explicit suffixes (e.g., current_A, power_W) so end-consumers of the data are unambiguously aware of units.

423-475: Optimize sanitizeMqttClientID.
The function handles character-by-character validations and transformations (lines 427-472). This logic works but can be simplified and made more maintainable by using a dictionary/map for replacements instead of large condition blocks. This reduces complexity and potential off-by-one errors.

365-413: Check for potential naming collisions in MQTT topics.
When creating Home Assistant sensors, the unique_id is built from mqttClientID and replaces spaces with hyphens. Ensure that collisions won’t occur if multiple devices share similar names or if character replacements become identical. Adding additional uniqueness factors like _i2cAddress might help.

424-476: Remove or clarify leftover commented-out code.
Multiple lines (e.g., 435, 438, 442, 446, 450, 454, 458, 459) contain references to extended replacements (“ue”, “oe”, “ss”) that are now commented out. Unless these are intentionally kept for later reference, removing them will reduce confusion and improve readability.

634-668: Guard against invalid intervals in config.
Your fallback handling in lines 643–647 helps avoid invalid intervals, but consider logging a message or storing an error code if check_interval is out of range. This ensures end users can diagnose misconfigurations quickly.

KrX3D commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,650 @@
+	** Function to publish sensor data to MQTT
+	**/
+	bool onMqttMessage(char* topic, char* payload) override {
+		if (!WLED_MQTT_CONNECTED) return false;
+		if (enabled) {
+			// Check if the message is for the correct topic
+			if (strcmp_P(topic, PSTR("/sensor/ina219")) == 0) {
```
example code of WLED:

      //if (strlen(topic) == 8 && strncmp_P(topic, PSTR("/command"), 8) == 0) {
      
      changing it to 
      
    
```
	char sensorTopic[128];
	snprintf_P(sensorTopic, sizeof(sensorTopic), "%s/sensor/ina219", mqttDeviceTopic);
	if (strcmp(topic, sensorTopic) == 0) {
```
	
	did not work, and it overwrote the "history" value that was stored in the mqtt topic

KrX3D committed:
coderabbitai ->

⚠️ Potential issue

Avoid casting float to bool for overflow checks.

The call to truncateDecimals(_ina219->getOverflow()) returns a float, which is then implicitly converted to bool. If the library returns non-integer values, this could cause unexpected results. For clarity and correctness, consider:

- overflow = truncateDecimals(_ina219->getOverflow());
+ overflow = _ina219->getOverflow() != 0;

KrX3D commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,650 @@
+			power = truncateDecimals(_ina219->getBusPower() / 1000.0); // Convert from mW to W
+			loadVoltage = truncateDecimals(busVoltage + (shuntVoltage / 1000));
+			overflow = truncateDecimals(_ina219->getOverflow());
+
```
done

KrX3D commented on usermods/INA219/usermod_ina219.h:
```diff
@@ -0,0 +1,650 @@
+		return sanitizedID;
+	}
+
+	/**
+	** Function to update energy calculations based on power and duration
+	**/
+	void updateEnergy(float power, unsigned long durationMs) {
+		float durationHours = durationMs / 3600000.0; // Milliseconds to hours
+		float energy_kWh = (power / 1000.0) * durationHours; // Watts to kilowatt-hours (kWh)
+		totalEnergy_kWh += energy_kWh; // Update total energy consumed
+
+		// Reset daily energy if 86400 seconds (24 hours) elapsed.
+		if (dailyResetTime >= 86400UL) {
+			dailyEnergy_kWh = 0;
+			dailyResetTime = 0;
+		}
+		dailyEnergy_kWh += energy_kWh;
+		dailyResetTime += durationMs / 1000;
+
+		// Reset monthly energy if 2592000 seconds (30 days) elapsed.
+		if (monthlyResetTime >= 2592000UL) {
+			monthlyEnergy_kWh = 0;
+			monthlyResetTime = 0;
+		}
+		monthlyEnergy_kWh += energy_kWh;
+		monthlyResetTime += durationMs / 1000;
+	}
```
values are stored in mqtt .> /sensor/ina219

after reconnecting it gets the values again, so it doesnt lose them and doesnt need to store them in RAM

KrX3D committed:
readme update, picture update

KrX3D committed:
changed .h to .cpp

KrX3D committed:
update usermod

KrX3D committed:
update usermod to v2

KrX3D committed:
usermod v2

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,665 @@
+
+	INA219_WE *_ina219 = nullptr; // INA219 sensor object
+
+	// Function to truncate decimals based on the configured decimal factor
+	float truncateDecimals(float val) {
+		if (_decimalFactor == 0) {
+			return roundf(val);
+		}
+		float factor = pow(10, _decimalFactor);
+		return roundf(val * factor) / factor;
+	}
```
Done

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,665 @@
+	/**
+	** Function to publish sensor data to MQTT
+	**/
+	bool onMqttMessage(char* topic, char* payload) override {
+		if (!WLED_MQTT_CONNECTED) return false;
+		if (enabled) {
+			// Check if the message is for the correct topic
+			if (strcmp_P(topic, PSTR("/sensor/ina219")) == 0) {
+				StaticJsonDocument<512> jsonDoc;
+
+				// Parse the JSON payload
+				DeserializationError error = deserializeJson(jsonDoc, payload);
+				if (error) {
+					return false;
+				}
```
Done

KrX3D committed:
coderabbitai improvments

KrX3D commented:
usermod was updated to v2, is working with the redone usermod PR.

KrX3D committed:
Update const.h

misspelled extension

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,661 @@
+	unsigned long lastCheck = 0; // Timestamp for the last check
+
+	// Define the variables using the pre-defined or default values
+	bool enabled = INA219_ENABLED;
+	uint8_t _i2cAddress = INA219_I2C_ADDRESS;
+	uint16_t _checkInterval = INA219_CHECK_INTERVAL;
+	uint16_t checkInterval = _checkInterval * 1000; // Convert seconds to milliseconds
+	INA219_ADC_MODE conversionTime = static_cast<INA219_ADC_MODE>(INA219_CONVERSION_TIME);
+	uint8_t _decimalFactor = INA219_DECIMAL_FACTOR;
```
Done

KrX3D committed:
coderabbitai improvments

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,663 @@
+					
+				// Serialize the JSON document into a character buffer
+				char buffer[1024];
+				size_t payload_size = serializeJson(jsonDoc, buffer);
```
Done

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,663 @@
+
+		// Serialize the JSON document into a temporary string
+		char buffer[1024];
+		size_t payload_size = serializeJson(doc, buffer);
```
Done

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,663 @@
+		configComplete &= getJsonValue(top["Enabled"], enabled);
+		configComplete &= getJsonValue(top[F("i2c_address")], _i2cAddress);
+
+		if (getJsonValue(top[F("check_interval")], checkInterval)) {
+			if (1 <= checkInterval && checkInterval <= 600) {
+				checkInterval *= 1000;
+			} else {
+				DEBUG_PRINTLN(F("INA219: Invalid check_interval value; using default."));
+				checkInterval = _checkInterval * 1000;
+			}
+		} else {
+			configComplete = false;
+		}
```
Partially done

KrX3D committed:
coderabbitai improvments

KrX3D committed:
daily/monthly kWh reset logic redone, resets now at midnight/1st day in month instead of 24h/30days

KrX3D committed:
coderabbitai improvments

DedeHai commented on usermods/INA219_v2/Readme.md:
```diff
@@ -0,0 +1,134 @@
+
+## Installation
+
+Add the compile-time option `-D USERMOD_INA219` to your `platformio.ini` (or `platformio_override.ini`) or use `#define USERMOD_INA219` in `my_config.h`.
```
this is outdated

KrX3D replied:
yeah you are right, i removed it.

DedeHai commented:
I am unable to test this, I can not even get it to show up in usermod menu for some reason.
does it play well with other I2C sensor usermods? Not sure how I2C is used in the library used.

is there a reason it is not compatible with INA219? the sensores seem to have almost the same functionalities.

KrX3D committed:
removed outdated installation methode

KrX3D commented:
> I am unable to test this, I can not even get it to show up in usermod menu for some reason. does it play well with other I2C sensor usermods? Not sure how I2C is used in the library used.

do you maybe have the same problem as here: 

https://github.com/wled/WLED/issues/4597

with the newest source code from git i only get it working with docker, on windows it failes with VS Code. but i get it working with the steps i mentioned here:

https://github.com/wled/WLED/issues/4597#issuecomment-2809975623

> is there a reason it is not compatible with INA219? the sensores seem to have almost the same functionalities.

not sure what you mean, do you mean the INA226 usermod? i didnt test it since i dont have that module, but it worked with some i2c temp sensors

KrX3D committed:
optimazations, more checks and log

KrX3D committed:
coderabbitai improvements

DedeHai commented:
> not sure what you mean, do you mean the INA226 usermod? i didnt test it since i dont have that module, but it worked with some i2c temp sensors

got the numbers mixed up, yes I mean INA226 which should work very much the same way, so would be good to support both using a #define since you already put in a lot of work and this UM is much more versatile than the old one.

KrX3D commented:
> got the numbers mixed up, yes I mean INA226 which should work very much the same way, so would be good to support both using a #define since you already put in a lot of work and this UM is much more versatile than the old one.

ok i will give it a try soon, but it will take some time, since i also want to buy an ina226 so i can try it out

KrX3D committed:
Fix for daily/monthly energy reset, wasnt working correctly before

DedeHai commented:
good question. what happens if you add in both libraries in the dependencies? does that eat up flash even if only one of them is used?
you could name it "INA2xx_Sensor" and if you have all the functionality (and more) of the original, the original could be removed to avoid confusion? Or is the new one incompatible with existing setups (MQTT topics etc)? If so, would it be easy to add a "compatibility" define so it is compatible to setups using the old one?  just a suggestion and a nice to have.

KrX3D commented:
> good question. what happens if you add in both libraries in the dependencies? does that eat up flash even if only one of them is used? 

ok i tried it with:

```
{
  "name": "INA219_v2",
  "dependencies": {
    "wollewald/INA219_WE":"~1.3.8",
    "wollewald/INA226_WE":"~1.2.9"
  }
}
```
and before and after the ram and flash size didnt change from:

```
RAM:   [==        ]  16.1% (used 52640 bytes from 327680 bytes)
Flash: [=======   ]  69.9% (used 1099014 bytes from 1572864 bytes)
```

the bin file is also the same size 1074kb, so i will keep it that way

>you could name it "INA2xx_Sensor" and if you have all the functionality (and more) of the original, the original could be removed to avoid confusion? 

ok i will rename it that way. i think when i add the ina226 the current once can be removed, since i also started with that usermod to create this one.

>Or is the new one incompatible with existing setups (MQTT topics etc)? If so, would it be easy to add a "compatibility" define so it is compatible to setups using the old one? just a suggestion and a nice to have.

hmm i need to look that up, but i think mostly the mqtt topics are the same. i think only the overflow changed since it wasnt working for me.
But when i fiished it i will compare it to the current mqtt topics

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,715 @@
+		sanitizedMqttClientID += "-" + String(escapedMac.c_str());
+
+		char sensorTopic[128];
+		snprintf_P(sensorTopic, 127, "homeassistant/%s/%s/%s/config", SensorType.c_str(), sanitizedMqttClientID.c_str(), sanitizedName.c_str());
+
```
done

KrX3D commented on usermods/INA219_v2/INA219_v2.cpp:
```diff
@@ -0,0 +1,715 @@
+		size_t payload_size = serializeJson(doc, buffer, sizeof(buffer));
+
+		char topic_S[128];
+		snprintf_P(topic_S, sizeof(topic_S), "homeassistant/%s/%s/%s/config", SensorType.c_str(), sanitizedMqttClientID.c_str(), sanitizedName.c_str());
+
```
done

KrX3D committed:
coderabbitai improvements

KrX3D committed:
- Generalized sensor naming from INA219 to INA2xx for broader compatibility
- Fixed issue with energy values being overwritten on startup or reflash
- Restored energy values from MQTT are now added to current totals instead of replacing them
- Added additional configuration options for INA sensor behavior

KrX3D committed:
renamed folder/file from ina219_v2 to ina2xx_v2

KrX3D commented on usermods/ina2xx_v2/Readme.md:
```diff
@@ -0,0 +1,130 @@
+# INA219 WLED Usermod
+
+This Usermod integrates the INA219 sensor with WLED to monitor energy consumption. It can read voltage, current, power, and calculate daily, monthly and total energy usage.
+
+## Features
+
+- Monitors bus voltage, shunt voltage, current, and power.
+- Calculates total energy consumed in kilowatt-hours (kWh).
+- Supports MQTT publishing of sensor data.
+- Publishes energy data to Home Assistant directly to the WLED entity for easy integration.
+- Displays daily, monthly and total energy used in the WLED GUI under the info section.
+- Configurable through WLED's web interface.
```
i will update this in the future when i add the ina226 part

KrX3D commented on usermods/ina2xx_v2/ina2xx_v2.cpp:
```diff
@@ -0,0 +1,1034 @@
+
+		if (!enabled) return true;
+
+		_logUsermodInaSensor("Creating new INA2xx instance with address 0x%02X", _i2cAddress);
+		_ina2xx = new INA219_WE(_i2cAddress);
+
+		if (!_ina2xx) {
+			_logUsermodInaSensor("Failed to allocate memory for INA2xx sensor!");
+			enabled = false;
+			return false;
+		}
```
i will update this in the future when i add the ina226 part

KrX3D committed:
coderabbitai improvements

KrX3D commented on usermods/ina2xx_v2/ina2xx_v2.cpp:
```diff
@@ -0,0 +1,948 @@
+		jsonDoc["monthlyResetTime"] = monthlyResetTime;
+
+		// Serialize the JSON document into a character buffer
+		char buffer[1024];
+		size_t payload_size = serializeJson(jsonDoc, buffer, sizeof(buffer));
```
done

KrX3D committed:
coderabbitai improvements

DedeHai commented:
what is the status?
I briefly checked the files, looks like the readme needs updating: UM is now called `ina2xx_v2` and it should also mention it supports INA226 as well and how to select which one to compile.

KrX3D commented:
Hi,
i received a ina226 and i already made changes but i couldnt totally test it so far since im on holiday. i think i get it done till monday and will update the files in git then

KrX3D committed:
fix "This branch has conflicts that must be resolved"

KrX3D committed:
Merge branch 'main' into 0_15

KrX3D committed:
Update pictures and readme

KrX3D committed:
update readme

KrX3D committed:
Fix numbering in Usage steps

KrX3D committed:
Update Usermod to support INA226

KrX3D commented:
Hi @DedeHai 

i updated the usermod to support the INA226.

I tested it in series with my Ina219 sensor which is running since months on a project and the values reported were nearly the same (0,002 difference)

I also updated the readme and some pictures

KrX3D committed:
updated picture and readme

KrX3D committed:
Merge branch 'wled:main' into 0_15

KrX3D committed:
### Fixed
- Fixed energy counter resets not working properly due to missing timestamp tracking
- Added validation to prevent energy calculation during invalid time periods (before NTP sync)
- Fixed missing `dailyResetTimestamp` and `monthlyResetTimestamp` initialization on first run
- Improved daily and monthly reset logic to use day/month comparison instead of hour/minute checks

### Added
- Added `dailyResetTimestamp` and `monthlyResetTimestamp` fields to track exact reset times
- Added timestamp fields to MQTT payload for Home Assistant `last_reset_value_template` support
- Added `state_class` configuration for Home Assistant sensors:
  - `total_increasing` for total energy (never resets)
  - `total` with `last_reset_value_template` for daily/monthly energy
  - `measurement` for instantaneous values (current, voltage, power)
- Added timestamp restoration from MQTT retained messages

KrX3D committed:
coderabbitai improvements

KrX3D committed:
coderabbitai improvements

KrX3D committed:
coderabbitai improvements

KrX3D committed:
Fix INA2xx MQTT energy restore

KrX3D committed:
Refine INA2xx MQTT publish and UI defaults

KrX3D committed:
Keep energy reset logic active at low power

KrX3D committed:
Adjust MQTT restore and reduce INA2xx log spam

KrX3D committed:
Fix INA2xx MQTT restore month lambda types

KrX3D committed:
Merge pull request #6 from KrX3D/codex/review-and-fix-ina2xx-usermod-implementation

Fix INA2xx MQTT energy restore handling

KrX3D committed:
Address INA2xx review feedback for MQTT and HA

KrX3D committed:
Allow INA226 conversion wait only during setup

KrX3D committed:
Restore INA226 conversion wait on all reinitializations

KrX3D committed:
Run INA226 conversion wait only during setup

KrX3D committed:
Self-heal future daily/monthly reset markers

KrX3D committed:
Reset energy counters on any day/month marker mismatch

KrX3D committed:
Merge pull request #21 from KrX3D/codex/review-and-fix-ina2xx-usermod-implementation

Address INA2xx PR feedback: MQTT topic safety, HA discovery, logging and INA226 init

softhack007 committed:
Merge branch 'main' into 0_15
